### PR TITLE
Fix request ID handling

### PR DESCRIPTION
**Pull request by sergefdrv**
This series combines fixing of various problems related to handling client request identifiers. These issues need to be solved to implement view change operation correctly. Summary of problems solved:
- Race condition when executing requests from different clients
- Handling of delayed reception of Request and Commit messages 
- Prepare/Commit message generation blocked by consensus rounds of previous requests
- Handling of duplicate Prepare messages and duplicate request execution attempts

### CODE REVIEW

File: core/internal/clientstate/client-state.go [hunk 1]
```diff
@@ -156,10 +128,8 @@ func (c *clientState) AddReply(reply *messages.Reply) error {
 	c.Lock()
 	defer c.Unlock()
 
-	if seq != c.lastAcceptedSeq {
-		return fmt.Errorf("request ID mismatch")
-	} else if seq == c.lastExecutedSeq {
-		return fmt.Errorf("duplicate Reply")
+	if seq <= c.lastExecutedSeq {
```
Comment (nhoriguchi) [inline core/internal/clientstate/client-state.go]: Currently request seq number is generated based on timestamp. So if an attacker creates a request message on the system with wrong clock and large sequence number is used, any messages generated based on proper timestamp could be discarded as old requests?

**Review by nhoriguchi — APPROVED** [review]

File: core/internal/clientstate/client-state.go (hunk 1 above)
Comment (sergefdrv) [reply core/internal/clientstate/client-state.go]: Please note that request identifiers (seq) are independent for each client. So if a compromised client exhausts its request ID space, it is screwed up. But only this client identity would be affected; others could continue normally. Although the consensus protocol cannot protect the client from being compromised, it provides safety and liveness when up to f replica nodes may be compromised (except tamperproof component).

Comment (nhoriguchi) [reply core/internal/clientstate/client-state.go]: OK, thank you for pointing out.

File: core/request.go [hunk 2]
```diff
@@ -101,28 +97,18 @@ func makeRequestHandler(id, n uint32, view viewProvider, verifier messageSignatu
...
-		// replicas when received a new request from a client.
-		// It must probably be done atomically with checking
-		// the request ID, otherwise there might be a race
-		// condition between checking the request ID and
-		// starting the timer here versus stopping the timer
-		// and supplying a Reply message when the request gets
-		// executed.
+		// TODO: A new request ID has arrived; the request
+		// timer should be re-/started in backup replicas at
+		// this point.
```
Comment (ynamiki) [inline core/request.go:133]: Does this timer mean the timer for a view change described in page 6 in [Efficient Byzantine Fault-Tolerance paper](https://www.researchgate.net/publication/260585535_Efficient_Byzantine_Fault-Tolerance) and #5?

Comment (sergefdrv) [reply core/request.go:133]: Yes, that refers to the request timer which triggers the view change operation.

**Pull request merged**